src/test/java/org/elasticsearch/NamingConventionTests.java

```diff
@@ -0,0 +1,101 @@
+            Path startPath = Paths.get(NamingConventionTests.class.getResource(path).toURI());
+
+            Files.walkFileTree(startPath, new FileVisitor<Path>() {
+                private Path pkgPrefix = Paths.get(path).getParent();
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                    pkgPrefix = pkgPrefix.resolve(dir.getFileName());
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    String filename = file.getFileName().toString();
+                    if (filename.endsWith("Tests.class") || filename.endsWith("Test.class")) { // don't worry about the ones that match the pattern
+                        return FileVisitResult.CONTINUE;
+                    }
+                    if (filename.endsWith(".class")) {
+                        try {
+                            if (Pattern.matches("[a-zA-Z]+[$].+\\.class", filename) == false) { // don't look at inner classes
+                                StringBuilder pkg = new StringBuilder();
+                                for (Path p : pkgPrefix) {
+                                    pkg.append(p.getFileName().toString()).append(".");
+                                }
+                                pkg.append(filename.substring(0, filename.length() - 6));
+
+                                Class<?> clazz = Class.forName(pkg.toString());
+                                if (ElasticsearchTestCase.class.isAssignableFrom(clazz)) {
+                                    if (Modifier.isAbstract(clazz.getModifiers()) == false) { // filter out abstract classes
+                                        classes.add(clazz);
+                                    }
+                                }
+                            }
+                        } catch (ClassNotFoundException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+                    throw exc;
+                }
+
+                @Override
+                public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+                    pkgPrefix = pkgPrefix.getParent();
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+
+        }
+        assertTrue("Not all subclasses of " + ElasticsearchTestCase.class.getSimpleName() +
+                " match the naming convention. Concrete classes must end with [Test|Tests]: " +  classes.toString(),
+                classes.isEmpty());
+    }
+
```

Comment:
the only case that we don't cover here is the classes that by mistake don't extend `ElasticsearchTestCase` (plain junit tests) whose name doesn't end with `Test` or `Tests` either..... pretty sure this never happened before. not sure we want to do anything about this. Maybe we could just separately check that there are no plain junit tests in the codebase.